packages/melody-streams/src/component.js

```diff
@@ -0,0 +1,137 @@
+    // part of the public API
+    this.el = element;
+    this.refs = {};
+    this.props = {};
+    this.propsStream = new BehaviorSubject({});
+    this.updates = new Subject();
+    this.subscriptions = [];
+    this.state = {};
+}
+
+Object.assign(Component.prototype, {
+    apply(props) {
+        this.propsStream.next(props);
+        if (this.subscriptions.length === 0) {
+            const t = this.getTransform({
+                dispatchCustomEvent(eventName, detail, options = {}) {
+                    const event = new CustomEvent(eventName, {
+                        ...options,
+                        detail,
+                    });
+                    this.el.dispatchEvent(event);
+                },
+                props: this.propsStream,
+                updates: this.updates,
+                subscribe: obs => this.subscriptions.push(obs.subscribe()),
+            });
+            const warningSubscription = warningTimer.subscribe();
+            const s = t
+                .pipe(
+                    distinctUntilChanged(shallowEqual),
+                    catchError(err => of(err))
+                )
+                .subscribe(
+                    state => {
+                        if (!warningSubscription.closed)
+                            warningSubscription.unsubscribe();
+                        this.state = state;
+                        enqueueComponent(this);
+                    },
+                    err => {
+                        if (process.env.NODE_ENV !== 'production') {
+                            /* eslint-disable no-console */
+                            console.error('Error: ', err);
+                            /* eslint-enable no-console */
+                        }
+                    }
+                );
+
+            this.subscriptions.push(s);
+        }
+    },
+
+    notify() {
+        this.updates.next();
+    },
+
+    componentWillUnmount() {
+        this.propsStream.complete();
+        this.updates.complete();
+        this.subscriptions.forEach(sub => sub.unsubscribe());
+        this.subscriptions.length = 0;
+    },
+
+    render() {},
+});
+
+const createComponentConstructor = Parent => {
+    function ChildComponent(el) {
+        Parent.call(this, el);
+    }
+    ChildComponent.prototype = Object.create(Parent.prototype, {
+        constructor: { value: ChildComponent },
+    });
+    return ChildComponent;
+};
+
+export const baseCreateComponent = (transform, templateFnOrObj) => {
+    const template = templateFnOrObj.render
+        ? props => templateFnOrObj.render(props)
+        : templateFnOrObj;
+    const ChildComponent = createComponentConstructor(Component);
+    ChildComponent.prototype.displayName =
+        transform.name || template.name || template.displayName || 'Unknown';
+    ChildComponent.prototype.render = function() {
+        return template(this.state);
+    };
+    ChildComponent.prototype.getTransform = transform;
```

Comment:
If we want to support cases where `transform` returns an object of streams we'll need to ensure that the return value is always wrapped in `combine`.

```suggestion
    ChildComponent.prototype.getTransform = api => combine(transform(api));
```

If the `transform` returns an observable, `combine` will simply return that. Otherwise it'll do its magic on the returned object.